nexus/src/app/sagas/instance_create.rs

```diff
@@ -535,85 +445,80 @@ async fn sic_allocate_instance_snat_ip(
         )
         .await
         .map_err(ActionError::action_failed)?;
-    Ok(())
+    Ok(ip_id)
 }
 
-/// Destroy all allocated _temporary_ external IPs (SNAT or Ephemeral)
-async fn sic_allocate_temporary_external_ips_undo(
+/// Destroy an allocated SNAT IP address for the instance.
+async fn sic_allocate_instance_snat_ip_undo(
     sagactx: ActionContext<SagaInstanceCreate>,
 ) -> Result<(), anyhow::Error> {
     let osagactx = sagactx.user_data();
     let datastore = osagactx.datastore();
     let saga_params = sagactx.saga_params();
     let opctx =
         OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
-    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+    let ip_id = sagactx.lookup::<Uuid>("snat_ip")?;
     datastore
-        .deallocate_instance_external_ip_by_instance_id(&opctx, instance_id)
+        .deallocate_instance_external_ip(&opctx, ip_id)
         .await
         .map_err(ActionError::action_failed)?;
     Ok(())
 }
 
-/// Create an ID for each external Ephemeral IP address.
-async fn sic_allocate_instance_ephemeral_ip_ids(
+/// Create an external IPs for the instance, using the request parameters at
+/// index `ip_index`, and return its ID if one is created (or None).
+async fn sic_allocate_instance_external_ip(
     sagactx: ActionContext<SagaInstanceCreate>,
-) -> Result<Vec<Uuid>, ActionError> {
+    ip_index: usize,
+) -> Result<Option<Uuid>, ActionError> {
+    let osagactx = sagactx.user_data();
+    let datastore = osagactx.datastore();
     let saga_params = sagactx.saga_params();
-    if saga_params.create_params.external_ips.len()
-        > MAX_EPHEMERAL_IPS_PER_INSTANCE as _
-    {
-        return Err(ActionError::action_failed(Error::invalid_request(
-            format!(
-                "At most {} Ephemeral IP(s) are supported",
-                MAX_EPHEMERAL_IPS_PER_INSTANCE
-            )
-            .as_str(),
-        )));
-    }
-    Ok(saga_params
-        .create_params
-        .external_ips
-        .iter()
-        .filter_map(|ip| {
-            if matches!(ip, params::ExternalIpCreate::Ephemeral { .. }) {
-                Some(Uuid::new_v4())
-            } else {
-                None
-            }
-        })
-        .collect())
+    let ip_params = saga_params.create_params.external_ips.get(ip_index);
+    let ip_params = match ip_params {
+        None => {
+            return Ok(None);
+        }
+        Some(ref prs) => prs,
+    };
+    let opctx =
+        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+
+    // Generate an ID and collect the possible pool name for this IP address
+    let ip_id = Uuid::new_v4();
```

Comment:
Same comment as earlier; the UUID generation should be distinct from actual allocations, otherwise the action is not idempotent - it would have the side effect of allocating distinct IPs